compiler/rustc_resolve/src/lib.rs

```diff
@@ -1584,6 +1586,57 @@ impl<'a> Resolver<'a> {
         });
     }
 
+    fn check_reexport_ambiguities(&mut self) {
+        for module in self.arenas.local_modules().iter() {
+            if let Some(_) = module.opt_def_id() {
+                let mut ambiguous_reexports = FxHashMap::default();
+
+                module.for_each_child(self, |this, ident, ns, binding| {
+                    if let NameBindingKind::Import { import: reexport, .. } = binding.kind
+                    && let res = binding.res().expect_non_local::<Res>()
+                    && res != def::Res::Err {
+                        if let Some((ambiguous_binding, _)) = binding.ambiguity
+                        && let NameBindingKind::Import { import: ambiguous_reexport, .. } = ambiguous_binding.kind {
+                            if let Some(did) = binding.res().opt_def_id()
+                            && let Some(local_did) = did.as_local()
+                            && let Some(amb_did) = ambiguous_binding.res().opt_def_id()
+                            && let Some(amb_local_did) = amb_did.as_local()
+                            {
+                                debug!("check_reexport_ambiguities: did = {:?}, amb_did = {:?}", did, amb_did);
+                                debug!("check_reexport_ambiguities: did.is_exported = {:?}, amb_did.is_exported = {:?}", this.effective_visibilities.is_exported(local_did), this.effective_visibilities.is_exported(amb_local_did));
+                                if this.effective_visibilities.is_exported(local_did)
+                                && this.effective_visibilities.is_exported(amb_local_did) {
+                                    ambiguous_reexports
+                                        .entry((ident, ns))
```

Comment:
I don't think grouping by `(iden, ns)` is necessary, you can report `buffer_lint_with_diagnostic` immediately here.